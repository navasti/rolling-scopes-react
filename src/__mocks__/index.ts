export { testOnChange, handleCloseModal } from './handlers';
export { localStorageMock } from './localStorage';

export {
  fetchBaseMock,
  fetchDetailsMock,
  handleResponseMock,
  fetchByParameterMock,
  handleMappedResponseMock,
} from './api';

export {
  links,
  pokemon,
  visible,
  invisible,
  isLoading,
  movesMock,
  typesMock,
  testErrors,
  pokemonData,
  pokemonMove,
  pokemonType,
  testingTitle,
  pokemonsMock,
  customPokemon,
  testingContent,
  testingLocation,
  testEmptyErrors,
  testingComponentName,
  moveModalContentProperties,
  typeModalContentProperties,
  pokemonModalContentProperties,
} from './data';

export {
  modalRef,
  errorRef,
  inputRef,
  selectRef,
  ModalTitle,
  ModalContent,
  TestingElement,
  TestingErrorElement,
} from './elements';
