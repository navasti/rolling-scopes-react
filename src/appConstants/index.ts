import { FormFields } from 'types';
import * as yup from 'yup';

export enum Limits {
  pokemon = 20,
  type = 10,
  move = 15,
}

export enum ErrorStatuses {
  success = '2',
  client = '4',
  server = '5',
}

export enum AvailableTabs {
  pokemons = 'pokemons',
  moves = 'moves',
  types = 'types',
}

export enum Fields {
  birthday = 'birthday',
  consent = 'consent',
  gender = 'gender',
  avatar = 'avatar',
  shiny = 'shiny',
  name = 'name',
  type = 'type',
}

export enum ErrorMessages {
  name = 'Name must be at least 2 characters long.',
  consent = 'Regulations consent is required.',
  birthday = 'Birthday is required.',
  type = 'Main type is required.',
  gender = 'Gender is required.',
}

export enum Types {
  fire = 'fire',
  grass = 'grass',
  poison = 'poison',
  flying = 'flying',
  ground = 'ground',
  psycho = 'psycho',
  water = 'water',
  ice = 'ice',
}

export const NONE = 'none';
export const MALE = 'male';
export const FEMALE = 'female';
export const INPUT_VALUE_KEY = 'input-value';
export const SUCCESS_MESSAGE = 'Pokemon added successfuly';
export const LINKS = ['home', 'about', 'form', 'invalid', 'details'] as const;
export const SEARCH_BAR_INSTRUCTIONS = [
  'Type and press enter to search for specific pokemon, type or move.',
  'Clear input and press enter to search for all pokemons, types and moves.',
] as const;

export const TABS = Object.values(AvailableTabs);
export const POKEMON_TYPES = Object.values(Types);

export const FORM_VALIDATION_SCHEMA = yup.object().shape({
  type: yup.string().required(ErrorMessages.type),
  name: yup.string().min(2, ErrorMessages.name).required(ErrorMessages.name),
  gender: yup.string().nullable().required(ErrorMessages.gender),
  avatar: yup
    .mixed()
    .test('fileSize', 'The file is too large', (value) =>
      !value?.length ? true : value?.[0]?.size <= 200000
    ),
  birthday: yup.string().required(ErrorMessages.birthday),
  consent: yup.bool().oneOf([true], ErrorMessages.consent),
  shiny: yup.bool(),
});

const BASE = 'https://pokeapi.co/api/v2';

export const API = {
  POKEMON_LIMIT: `?limit=${Limits.pokemon}`,
  TYPE_LIMIT: `?limit=${Limits.type}`,
  MOVE_LIMIT: `?limit=${Limits.move}`,
  POKEMON: `${BASE}/pokemon`,
  TYPE: `${BASE}/type`,
  MOVE: `${BASE}/move`,
  BASE,
};

export const FIELDS_VALIDATION_BY_NAME = {
  TEXT: [Fields.name, Fields.birthday, Fields.type],
  CHECK: [Fields.consent, Fields.gender],
};

export const DEFAULT_VALUES: FormFields = {
  consent: false,
  avatar: null,
  gender: MALE,
  shiny: false,
  birthday: '',
  name: '',
  type: '',
};
