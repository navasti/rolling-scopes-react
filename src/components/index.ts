export { Loader } from './Loader';
export { Modal } from './Modal';
