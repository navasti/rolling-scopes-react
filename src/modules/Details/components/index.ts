export { MoveCardDetails } from './MoveCardDetails';
export { TypeCardDetails } from './TypeCardDetails';
export { PokemonCardDetails } from './PokemonCardDetails';
