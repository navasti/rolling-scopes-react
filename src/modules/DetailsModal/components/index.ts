export { TypeModalContent } from './TypeModalContent';
export { MoveModalContent } from './MoveModalContent';
export { PokemonModalContent } from './PokemonModalContent';
