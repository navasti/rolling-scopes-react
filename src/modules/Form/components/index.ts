export { BirthdayField } from './BirthdateField';
export { ErrorMessage } from './ErrorMessage';
export { ConsentField } from './ConsentField';
export { AvatarField } from './AvatarField';
export { GenderField } from './GenderField';
export { ShinyField } from './ShinyField';
export { TypeField } from './TypeField';
export { NameField } from './NameField';
export { FormCard } from './FormCard';
