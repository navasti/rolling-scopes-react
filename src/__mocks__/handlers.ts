import { screen } from '@testing-library/react';
import { AvailableTabs } from 'appConstants';

export const expectByText = jest.fn((text: string) => {
  expect(screen.getByText(text)).toBeInTheDocument();
});

export const handleCloseModal = jest.fn();
export const handleCardSelect = jest.fn();
export const handleOpenModal = jest.fn();
export const onTabClick = jest.fn((availableTab: AvailableTabs) => availableTab);
