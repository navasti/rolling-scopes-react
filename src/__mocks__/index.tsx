import { ErrorMessages } from 'appConstants';
import { createRef, RefObject } from 'react';
import { hasError } from 'utils';
import {
  MockDetailedDataType,
  PokemonMoveDetails,
  PokemonTypeDetails,
  MockBaseDataType,
  PokemonDetails,
  CustomPokemon,
  ErrorsObject,
  CheckFields,
  PokemonData,
  PokemonMove,
  PokemonType,
  TextFields,
  Pokemon,
} from 'types';

export const visible = true;
export const invisible = false;
export const isLoading = false;

export const errorRef: RefObject<HTMLSpanElement> = createRef();
export const inputRef: RefObject<HTMLInputElement> = createRef();
export const selectRef: RefObject<HTMLSelectElement> = createRef();

export const links = ['Home', 'About', 'Form', 'Invalid'];
export const componentName = 'About';
export const location = '/about';
export const testingContent = 'Content';
export const testingTitle = 'Title';

export const TestingElement = () => <p>testing element</p>;
export const TestingErrorElement = () => <p>testing error</p>;

export const TestingTitleElement = () => <span>{testingTitle}</span>;
export const TestingContentElement = () => <div>{testingContent}</div>;
export const modalRef: RefObject<HTMLDivElement> = createRef();

export const fetchBaseMock = <T extends MockBaseDataType>(mockData: T) => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      status: 200,
      json: () => Promise.resolve({ results: [mockData] }),
    })
  ) as jest.Mock;
};

export const fetchByParameterMock = <T extends MockDetailedDataType>(mockData: T) => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      status: 200,
      json: () => Promise.resolve(mockData),
    })
  ) as jest.Mock;
};

export const fetchDetailsMock = <T,>(mockData: Array<T>) => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      status: 200,
      json: () => Promise.resolve(mockData[0]),
    })
  ) as jest.Mock;
};

export const pokemonType: PokemonType = {
  name: 'typex',
  url: 'url',
};

export const pokemon: Pokemon = {
  name: 'pokemonix',
  url: 'url',
};

export const pokemonMove: PokemonMove = {
  name: 'movix',
  url: 'url',
};

export const typeModalContentProperties = [
  'Double damage from',
  'Double damage to',
  'Half damage from',
  'Half damage to',
  'No damage from',
  'No damage to',
  'Pokemons',
  'Moves',
] as const;

export const moveModalContentProperties = [
  'Learned by',
  'Priority',
  'Accuracy',
  'Power',
  'Type',
  'PP',
] as const;

export const pokemonModalContentProperties = [
  'Pokemon statistics',
  'Pokemon abilities',
  'Base experience',
  'Height',
  'Weight',
  'Types',
] as const;

export const pokemonData: PokemonData = {
  results: [{ name: 'testName', url: 'testURL' }],
  previous: null,
  next: null,
  count: 1,
};

export const testOnChange = (field: TextFields | CheckFields) => console.log(field);

export const testEmptyErrors: ErrorsObject = {
  birthday: null,
  consent: null,
  gender: null,
  name: null,
  type: null,
};

export const testErrors: ErrorsObject = {
  birthday: ErrorMessages.birthday,
  consent: ErrorMessages.consent,
  gender: ErrorMessages.gender,
  name: ErrorMessages.name,
  type: ErrorMessages.type,
};

export const customPokemon: CustomPokemon = {
  birthday: '01-01-1999',
  name: 'pokemonix',
  gender: 'male',
  avatar: null,
  type: 'fire',
  shiny: true,
};

export const detailedTypes: Array<PokemonTypeDetails> = [
  {
    damage_relations: {
      double_damage_from: [],
      double_damage_to: [{ name: 'toTest', url: 'urlTest' }],
      half_damage_from: [{ name: 'fromTest', url: 'urlTest' }],
      half_damage_to: [],
      no_damage_from: [],
      no_damage_to: [],
    },
    pokemon: [{ pokemon: { name: 'pokemonTest', url: 'urlTest' }, slot: 1 }],
    moves: [{ name: 'moveTest', url: 'urlTest' }],
    name: 'typeTest',
    id: 10,
  },
];

export const detailedMoves: Array<PokemonMoveDetails> = [
  {
    learned_by_pokemon: [{ name: 'pokomonTest', url: 'urlTest' }],
    type: { name: 'typeTest', url: 'urlTest' },
    name: 'moveTest',
    accuracy: 60,
    priority: 0,
    power: 50,
    id: 20,
    pp: 15,
  },
];

export const detailedPokemons: Array<PokemonDetails> = [
  {
    abilities: [{ ability: { name: 'ability' }, is_hidden: false, slot: 2 }],
    stats: [{ base_stat: 33, effort: 33, stat: { name: 'test' } }],
    sprites: { back_default: 'url', front_default: 'url' },
    types: [{ slot: 2, type: { name: 'typeTest' } }],
    name: 'Testing pokemon',
    base_experience: 12,
    height: 15,
    weight: 10,
    id: 34,
  },
  {
    abilities: [{ ability: { name: 'ability2' }, is_hidden: true, slot: 1 }],
    stats: [{ base_stat: 22, effort: 22, stat: { name: 'test' } }],
    sprites: { back_default: 'url2', front_default: 'url2' },
    types: [{ slot: 2, type: { name: 'typeTest2' } }],
    name: 'Testing pokemon2',
    base_experience: 16,
    height: 11,
    weight: 11,
    id: 35,
  },
];

type Store = Record<string, string>;

export const localStorageMock = (() => {
  let store: Store = {};
  return {
    getItem(key: string) {
      return store[key];
    },
    setItem(key: string, value: string) {
      store[key] = value.toString();
    },
    clear() {
      store = {};
    },
    removeItem(key: string) {
      delete store[key];
    },
  };
})();
