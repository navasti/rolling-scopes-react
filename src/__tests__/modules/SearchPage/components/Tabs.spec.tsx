import { screen, render, fireEvent } from '@testing-library/react';
import { expectByText, onTabClick } from '__mocks__/handlers';
import { Tabs } from 'modules/SearchPage/components';
import { AvailableTabs } from 'appConstants';
import { TABS } from 'appConstants';

describe.skip('Tabs', () => {
  afterEach(() => expectByText.mockReset());
  it('tabs should match snapshot and display lengths if isLoading prop equal false', () => {
    const { container } = render(
      <Tabs
        activeTab={AvailableTabs.pokemons}
        onClick={onTabClick}
        lengths={lengths}
        isLoading={false}
        options={TABS}
      />
    );
    for (const length in lengths) {
      const key = length as keyof typeof lengths;
      TABS.forEach((tab) => tab === key && expectByText(`${tab} (${lengths[key]})`));
    }
    expect(expectByText).toHaveBeenCalledTimes(TABS.length);
    expect(container).toMatchSnapshot();
  });
  it('tabs should match snapshot and should not display lengths if isLoading prop equals true', () => {
    const { container } = render(
      <Tabs
        activeTab={AvailableTabs.pokemons}
        onClick={onTabClick}
        lengths={lengths}
        isLoading={true}
        options={TABS}
      />
    );
    TABS.forEach((tab) => expect(screen.getByText(tab)).toBeInTheDocument());
    expect(container).toMatchSnapshot();
  });
  it('clicking on tab should invoke onClick method', () => {
    const { container } = render(
      <Tabs
        activeTab={AvailableTabs.pokemons}
        onClick={onTabClick}
        lengths={lengths}
        isLoading={true}
        options={TABS}
      />
    );
    const spans = container.querySelectorAll('span');
    spans.forEach((span) => {
      fireEvent.click(span);
      expect(onTabClick).toBeCalledWith(span.textContent?.trim());
    });
    expect(onTabClick).toBeCalledTimes(spans.length);
    expect(spans.length).toEqual(TABS.length);
  });
});
