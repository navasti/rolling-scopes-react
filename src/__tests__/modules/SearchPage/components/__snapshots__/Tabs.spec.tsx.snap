// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs tabs should match snapshot and display lengths if isLoading prop equal false 1`] = `
<div>
  <div
    class="sc-olbas ghNiGT"
  >
    <span
      class="sc-hiMGwR hAxTaP"
    >
      pokemons
       
      (10)
    </span>
    <span
      class="sc-hiMGwR knipss"
    >
      moves
       
      (15)
    </span>
    <span
      class="sc-hiMGwR knipss"
    >
      types
       
      (20)
    </span>
  </div>
</div>
`;

exports[`Tabs tabs should match snapshot and should not display lengths if isLoading prop equals true 1`] = `
<div>
  <div
    class="sc-olbas ghNiGT"
  >
    <span
      class="sc-hiMGwR hAxTaP"
    >
      pokemons
       
    </span>
    <span
      class="sc-hiMGwR knipss"
    >
      moves
       
    </span>
    <span
      class="sc-hiMGwR knipss"
    >
      types
       
    </span>
  </div>
</div>
`;
